### PR TITLE
[Dispatch] Add pattern to bubble expand through extract 1/2

### PR DESCRIPTION
**Pull request by IanWood1**
This is the 1/2 changes needed to reland https://github.com/iree-org/iree/pull/18857 (with an open PR https://github.com/iree-org/iree/pull/19113).

Adds pattern to bubble up expand shape through extract slice. i.e `expand(extract)` to `extract(expand)`. This only supports the case where the expanded dimensions are not modified by the extract slice and there are no dynamic dimensions.

This is important because `tensor.expand_shape` ops _cannot be cloned_  while `tensor.extract_slice` ops _can be cloned_. So, if the `expand_shape` gets stuck on the bottom of the `extract_slice` it will block it from being cloned and the `extract_slice` will have to be put into its own dispatch.

### CODE REVIEW

File: compiler/src/iree/compiler/DispatchCreation/test/bubble_up_expand_shapes.mlir [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  util.return %expanded_239 : tensor<2x64x64x2560xf16>
+}
+
+// CHECK-LABEL:  @bubbble_expand_through_extract
+//       CHECK:    %[[EXPAND:.+]] = tensor.expand_shape
+//       CHECK:    %[[EXTRACT:.+]] = tensor.extract_slice %[[EXPAND]]
+
+// -----
+
+
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/test/bubble_up_expand_shapes.mlir]: nit: remove one blank line

File: compiler/src/iree/compiler/DispatchCreation/test/bubble_up_expand_shapes.mlir [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// RUN: iree-opt --split-input-file --mlir-print-local-scope --pass-pipeline="builtin.module(util.func(iree-dispatch-creation-bubble-up-expand-shapes, canonicalize, cse, canonicalize))" %s | FileCheck %s
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/test/bubble_up_expand_shapes.mlir]: Do we need to run `canonicalize, cse, canonicalize` in the test? Can we drop it?

File: compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp [hunk 3]
```diff
@@ -38,6 +38,77 @@ struct BubbleUpExpandShapesPass final
...
+  using OpRewritePattern<tensor::ExpandShapeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExpandShapeOp expandOp,
+                                PatternRewriter &rewriter) const override {
+    auto extractOp = expandOp.getSrc().getDefiningOp<tensor::ExtractSliceOp>();
+    if (!extractOp) {
+      return failure();
+    }
+
+    auto origType = extractOp.getSourceType();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp]: I think using `srcType` (which means that it is the source tensor type) is easier to follow.

File: compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp [hunk 4]
```diff
@@ -38,6 +38,77 @@ struct BubbleUpExpandShapesPass final
   void runOnOperation() override;
 };
 
+// TODO: move this upstream with other tensor bubbling patterns.
+struct BubbleExpandThroughExtract final
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp:46]: nit: add a doc to the pattern.

File: compiler/src/iree/compiler/DispatchCreation/test/bubble_up_expand_shapes.mlir (hunk 2 above)
Comment (IanWood1) [reply compiler/src/iree/compiler/DispatchCreation/test/bubble_up_expand_shapes.mlir]: Nope, doesn't look needed. dropped it

File: compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp [hunk 5]
```diff
@@ -38,7 +38,10 @@ struct BubbleUpExpandShapesPass final
   void runOnOperation() override;
 };
 
-// TODO: move this upstream with other tensor bubbling patterns.
+/// Bubbles a `tensor.expand_shape` op through a `tensor.extract_slice` op. This
+/// pattern only gets applied when the `extract_slice` doesn't modify dimensions
+/// that are expanded by the `expand_shape`.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp]: IIRC, it only works for static shapes? Dynamic shapes could be very tricky, and I think we're not going to support that in the near future. So we want to document that it only works for static shapes.

**Review by hanhanW — APPROVED** [review]
LGTM, just one nit about the doc.

File: compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp (hunk 5 above)
Comment (IanWood1) [reply compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp]: Yeah, only static shapes so I updated the comment to reflect that

**Pull request merged**